---
description: Docs intro
layout: ../../../layouts/MainLayout.astro
---

# Python 异步编程

## 为什么要异步？

我们先来理解异步的概念。**异步(async)**是相对于**同步(sync)**而言的。

**同步**就是一件事一件事的执行，遇到具有IO耗时任务时（如网络请求，`time.sleep()`等需要程序等待的任务），它会不断等待，直到任务完成后，才继续进行后面的任务；

而在**异步**环境下，进程不必等待请求或其他阻塞执行的代码块（耗时操作）执行完毕，即可继续执行后面的代码。耗时任务会被挂起，交给**协程**处理。

由此可见，在一些情景下，异步的效率是要远远高于同步的，因为它省去了大部分等待的时间，更好地发挥了主机的性能。

自从Python3内部支持了异步模块`asyncio`模块后，Python业内的一些框架（如`tornado`、`fastapi`（性能极佳）、 `django 3.x asgi`包括`django channels`）或是模块（如`aiohttp`）都在往**异步非阻塞**方向发展，为了提升性能。异步是大势所趋。

## 基于协程的异步模块

在Python 3.4及以上的版本，Python推出了内置的`asyncio`模块，实现了异步。

异步与协程密切相关，上一节中留了一个悬念，即可以使用异步模块实现协程，写法如下：

```python
import asyncio

@asyncio.coroutine
def func1():
    print("1")
    yield from asyncio.sleep(2)
    print("2")

@asyncio.coroutine
def func2():
    print("3")
    yield from asyncio.sleep(2)
    print("4")

tasks = [
    asyncio.ensure_future(func1()),
    asyncio.ensure_future(func2())
]

loop = asyncio.get_event_loop()
loop.run_until_complete(asyncio.wait(tasks))
```

代码看不懂暂时没有关系，我们来分析它是如何实现协程的：

+ 首先，`asyncio`模块中的装饰器`@asyncio.coroutine`，可以将一个普通函数转换为一个协程函数，上方的`func1`和`func2`两个函数即为协程函数。

  协程函数无法通过`函数名()`的形式被执行，必须通过`asyncio.get_event_loop().run_until_complete(asyncio.wait(list))`执行，需要传递一个`future`对象列表`list`，这个和`asyncio.ensure_future`封装后面会详细讲。

  由于是一个线程在执行任务，故列表中任务执行顺序随机。

+ 当协程函数执行过程中遇到了`IO`耗时操作（如上面样例中的`asyncio.sleep(2)`），语法上，要添加`yield from`关键字；执行上，程序会**自动地**切换到任务列表中，并执行其它未执行完毕的协程函数，而该耗时操作则会被挂起。

  对上方的代码可以这样解释，加入先执行`func1`，打印完`1`后，遇到了耗时操作，此时不会等待，而是搜索到任务列表中有未执行的任务`func2`，于是执行`func2`，直到`func2`中也遇到了耗时操作；当`func1`中的耗时操作结束后，执行完`func1`，再执行完任务列表中的`func2`，全部任务完成。

相较于上节的`greenlet`和`gevent`，`asyncio`的优势在于：遇到耗时操作，后者会**自动切换**到其它任务，而非前者的手动操作。这就实现了**异步**——无需等待阻塞操作，直接执行其它任务。

在Python 3.5后，引入了`async`和`await`关键字，替代了`@asyncio.coroutine`的装饰器写法，新版本的写法（推荐）如下：

```python
import asyncio

async def func1():
    print("1")
    await asyncio.sleep(2)
    print("2")

async def func2():
    print("3")
    await asyncio.sleep(2)
    print("4")

tasks = [
    asyncio.ensure_future(func1()),
    asyncio.ensure_future(func2())
] 

loop = asyncio.get_event_loop()
loop.run_until_complete(asyncio.wait(tasks))
```

即删去原先的装饰器，在`def`关键字前添加`async`，耗时操作前的`yield from`被替换为`await`关键字。

仅仅是写法改变了，实际执行情况并不变。

## `asyncio`异步编程

### 事件循环

事件循环可以理解成一个**检测并执行**某些代码的**死循环**，某些代码均会被储存在**任务列表**中，执行流程可以抽象为：

```python
任务列表 = [任务1, 任务2, 任务3, ...]
while True:
    可执行的任务列表, 已完成的任务列表 = 去任务列表中检查所有的任务，将'可执行'和'已完成'的任务返回
    
    for 就绪任务 in 已经就绪的任务列表:
        执行已就绪的任务
        
    for 已完成的任务 in 已完成的任务列表:
        在任务列表中移除已完成的任务
    
    if 任务列表中的任务全部完成
    	终止循环
```

可以通过`asyncio`模块中的`get_event_loop()`函数生成或获取一个事件循环(对象)：

```python
import asyncio
# 生成或获取一个事件循环
loop = asyncio.get_event_loop()
```

通过事件循环对象的`run_until_complete(tasks)`方法，传入一个任务`tasks`，可将任务`tasks`放到任务列表中：

```python
# 将任务tasks放到任务列表中
loop.run_until_complete(tasks)
```

`get_event_loop()`还有一个替代函数`get_running_loop()`，区别是前者可以在任何环境下使用，后者只能在协程环境中（即`async`修饰的函数内）使用，否则会抛出`RuntimeError`异常。

### 执行协程任务

 我们在上面提到，带`async`关键字修饰的函数会变为协程函数，执行`函数名()`会返回一个**携程对象**，而不会执行函数体的内容。

要想执行携程函数体的内容，必须将其交给事件循环来处理：

```python
import asyncio

async def func():
    print('执行协程函数体')

cor_obj = func()    # 创建协程对象
print(cor_obj)      # <coroutine object func at 0x000001913BF4D2C0>

loop = asyncio.get_event_loop()
loop.run_until_complete(cor_obj)    # 传入携程对象，执行协程函数体
```

在Python 3.7+ 的版本，我们有了更简洁的函数`asynio.run()`，上面代码最后两行等价于：

```python
asynio.run(cor_obj)
```

`asynio.run()` 只是一种简写，执行的具体流程与原写法一致。

### `await` 关键字

我们刚才提到`await`关键字写在耗时操作之前，这里的"耗时操作"其实是一种**可等待的对象**。

写在`await`的**可等待的对象**分为三种：协程对象，`Future`对象、`Task`对象，目前将它们理解为`IO`等待即可。

`Future`对象和`Task`对象我们放到下面讲，先看一段`await + 协程对象`的示例：

```python
import asyncio

async def others():
    print('start')
    await asyncio.sleep(2)
    print('end')
    return "success"

async def func():
    print("Coroutine code is running")
    resp1 = await others()	# await + 协程对象
    print('IO request ends, response is', resp1)
    resp2 = await others()
    print('IO request ends, response is', resp2)

asyncio.run(func())

''' 执行结果：
Coroutine code is running
start
end
IO request ends, response is success
start
end
IO request ends, response is success
'''
```

解释：通过`run`执行协程函数`func`后，遇到了一个`await`，后面跟了一个协程对象`others()`，此时不会立即切换，而是进入`others`中执行其函数体，直到遇到耗时操作，再进行切换。

这里看似执行结果与同步没有差异，其实是因为任务列表中没有其他任务，程序线性地执行，于是出现了这样的结果。

### `Task` 对象

下面我们聊聊`Task`对象。

简单地来说，它是用来**在事件循环中添加多个任务**。

可使用`asyncio.create_task()`或`asyncio.ensure_future()`函数创建一个任务对象，需要传入一个可等待的对象。

它们的区别是，`asyncio.create_task()`更为底层，只支持传入协程对象；而`asyncio.ensure_future()`对`asyncio.create_task()`做了封装，支持传入协程对象和`Future`对象，更建议使用`asyncio.ensure_future())`。

示例：

```python
import asyncio

async def func():
    print(1)
    await asyncio.sleep(2)
    print(2)
    return 'success'

async def main():
    print('main start')
    # 创建一个Task对象，将当前执行func函数任务添加到事件循环
    # 在传入协程对象的场景，与create_task()函数作用相同
    task1 = asyncio.ensure_future(func())
    task2 = asyncio.ensure_future(func())
    print('main end')
    # 当执行某协程遇到IO操作时，会自动切换到其他协程
    res1 = await task1
    res2 = await task2
    print(res1, res2)

asyncio.run(main())
''' 输出结果
main start
main end
1
1
2
2
success success
'''
```

通过`create_task()`函数在任务列表中添加了两个协程对象，于是在`await task1`时，执行`func`中的`await asyncio.sleep(2)`，会切换到任务列表中未执行完毕的`task2`中执行，因此打印结果有两个连续的`1`。

上方的代码写法了解即可，一般会将代码这样优化，执行逻辑不变：

```python
...

async def main():
    print('main start')

    task_list = [
        asyncio.ensure_future(func()),
        asyncio.ensure_future(func())
    ]
    print('main end')
    # 当执行某协程遇到IO操作时，会自动切换到其他协程
    done, pending = await asyncio.wait(task_list)
    print(done)
    print(done.pop().result())

asyncio.run(main())
```

我们可以讲创建的`Task`对象放入一个列表中，这样只用一个`await`即可启动所有事件循环。

由于`await`后面只能跟可等待的对象（列表不行），这里需要使用`asyncio.wait()`函数将列表转换为协程对象。该函数执行完毕后会返回一个元组，其中包含两个元素`done`和`pending`。

+ `done`中包含一个**集合**，元素即为原列表中的`Task`对象，其中的`result`成员变为了对应协程函数的返回值，可通过`done.pop().result()`获取返回值。

+ `pending`用的比较少。`asyncio.wait()`函数不仅可以传入任务列表，还可以传入`timeout`形参，表示等待的最长时间（单位为秒），`pending`为一个集合，元素为未在这段事件内完成的`Task`对象，若不指定`timeout`则会无限等待。

  可将上面的第`12`行改为：`done, pending = await asyncio.wait(task_list, timeout=1)`，打印`pendding`尝试一下。由于只等`1`秒，而两个任务都需要`2`秒才能完成，故此时`pending`中包含两个对象。

要注意的是，如果写法改为：

```python
import asyncio

async def func():
    print(1)
    await asyncio.sleep(2)
    print(2)
    return 'success'

task_list = [
    asyncio.ensure_future(func()),
    asyncio.ensure_future(func())
]

done, pending = asyncio.run(asyncio.wait(task_list))
```

此时程序会报错。乍一看只是将`main`函数的内容拆分出来了，为什么会有问题呢？

其实是因为在`main`函数封装的版本，调用`create_task`之前，在`run`方法内部已经创建了事件循环；而后者在未创建事件循环的情况下就调用了`create_task`，肯定会报错。

解决方法也很简单，将`asyncio.create_task()`删去即可：

```python
...
task_list = [
    func(),
    func()
]
...
```

在`run()`方法生成事件循环后，会自动将协程对象列表封装成`Task`对象列表，进而正常地启动事件循环。

### `asyncio.Future` 对象

`Future`类是`Task`类的父类，`Task`对象内部`await`**结果的处理**是基于`Future`对象的。

我们来看一个示例：

```python
import asyncio

async def set_after(fut: asyncio.Future):   # 此处的类型标注是可选的，只是为了代码提示
    await asyncio.sleep(2)
    fut.set_result("666")

async def main():
    # 获取当前事件循环
    loop = asyncio.get_event_loop()
    # 创建一个任务（Future对象），没有绑定任何行为，则这个任务永远不知道什么时候结束
    fut = loop.create_future()
    # 创建一个任务（Task对象），绑定了set_after函数，函数内部在2s之后，会给fut赋值
    # 即手动设置future任务的最终结果，那么fut就可以结束了
    await loop.ensure_future(set_after(fut))
    # 等待Future对象获取最终结果，如果不设置最终结果，那么这里会一直阻塞
    data = await fut
    print(data)

asyncio.run(main())
```

在这个示例中，使用了`loop`对象的`create_future`方法创建了一个`Future`对象`fut`，此时创建的`fut`对象没有绑定任何行为。若不给其指定行为或结果，程序会一直阻塞。

这里在`set_after()`函数中，使用`Future`对象的`set_result`方法可以手动设置`Future`对象的返回结果。设置了结果后，`await fut`就能正常返回了。

`Future`对象更接近底层，几乎不需要在一般代码中体现。

## 异步迭代器

在之前的小节中，我们知道可迭代对象即具备了`__iter__`方法的对象；迭代器即一个实现了`__iter__()`方法和`__next__()`方法的对象。

类似地，有这样的定义：

+ 一个实现了`__aiter__()`方法的对象，为**异步可迭代对象**；
+ 一个实现了`__aiter__()`方法和`__anext__()`方法的对象，为**异步迭代器**；

先看下面的一个定义异步迭代器的示例：

```python
import asyncio

class Counter:
    """ 自定义异步迭代器（同时也是异步可迭代对象） """
    def __init__(self):
        self.num = 0

    async def count(self):
        # await asyncio.sleep(1)
        self.num += 1
        if self.num > 10:
            return None
        return self.num

    def __aiter__(self):
        return self

    async def __anext__(self):
        val = await self.count()
        if val is None:
            raise StopAsyncIteration
        return val
```

其中`__aiter__()`方法需要返回一个异步可迭代对象，一般`return self`就行了；

`__anext__()`应当为一个协程方法，需要使用`async`修饰，返回一个协程对象，因此上方的`readline`即使没有耗时操作也需要`async`修饰。

迭代器中定义的`StopIteration`异常，这里更换为了`StopAsyncIteration`异常。

可使用`async for`循环遍历异步迭代器，语法同`for`：

```python
async for i in async_iterator:
	print(i)
```

但按照语法要求，`async for`循环**必须写在协程函数内**，遍历上方迭代器`Counter`示例：

```python
async def main():
    c_obj = Counter()
    async for i in c_obj:
        print(i)

asyncio.run(main())
```

## 上下文管理与异步上下文管理

### `with`上下文管理

在之前的[文件操作](https://docs.drshw.tech/pb/primary/12/)一节中，我们已经用到了`with`和`as`关键字，那它到底是如何进行运作的呢？

`with`关键字后面必须要跟一个**上下文管理器对象**，该对象中必须实现两个方法——`__enter__()`方法和`__exit__()`方法：

`with`的执行流程如下：

- `with`语句执行时，先获取上下文管理器对象，随后调用其 `__enter__()`方法：

  - `__enter__()`方法可以返回上下文管理器对象本身，也可以是其他相关对象；

- 若有 `as var` 语句，则将 `__enter__()`方法返回值赋给变量`var`；

  例如`with open('123.txt', 'wt') as f:`，会先执行`open()`，实例化出一个`File`对象，再调用该对象的`__enter__()`方法，再将其赋值给变量`f`。

- 当代码块执行完成时，调用`__exit__()`方法：

  + `__exit__()`带有三个参数 `exc_type`,，`exc_val`，`exc_tb`；
  + 若上下文管理器对象执行无异常，则三个参数均为 `None`；
  + 若发生异常，则三个参数分别为 异常类型，异常值和tracback信息；

其实之所以使用`with`语句后，文件就不需要手动关闭，是因为在`File`对象中定义了`__exit__()`方法，用于关闭文件。

了解了执行逻辑后，我们尝试编写一个文件上下文管理类`FileManager`：

```python
class FileManager:
    def __init__(self, filename, mode):
        print("调用__init__方法")
        self.filename = filename
        self.mode = mode
        self.file = None
	
    # 编写enter方法，在进入with代码块前调用
    def __enter__(self):
        print("调用__enter__方法")
        self.file = open(self.filename, self.mode)
        return self.file

    # 执行完with代码块后调用，直接关闭文件
    def __exit__(self, exc_type, exc_val, exc_tb):
        print("调用__exit__方法")
        if self.file:
            self.file.close()

# 测试
with FileManager("test.txt", "w") as f:
    print("调用with代码块")
    f.write("hello world")

''' 执行结果
调用__init__方法
调用__enter__方法
调用with代码块
调用__exit__方法
'''
```

### 异步上下文管理

在上面我们已经了解到，上下文管理器对象需要在类内实现`__enter__()`方法和`__exit__()`方法。

类似地，有这样的要求：

+ 异步上下文管理器对象需要在类内实现`__aenter__()`方法和`__aexit__()`方法，控制`async with`语句中的环境。

其余语法和执行流程与普通上下文管理`with ... as ...`一致，但要注意的是`__aenter__()`方法和`__aexit__()`方法需要`async`，且按照语法规则，`async with`语句必须写在协程函数内，我们直接看一个示例：

```python
import asyncio

class AsyncioContextManager:
    def __init__(self, vars):
        self.vars = vars
        
	# 两个魔术方法都需要async修饰
    async def __aenter__(self):
        print("调用__aenter__方法")
        return self

    async def __aexit__(self, exc_type, exc, tb):
        print("调用__aexit__方法")
        await asyncio.sleep(1)

    async def do_something(self):
        return self.vars

async def main():
    # async with需要放在协程函数内
    async with AsyncioContextManager("test") as acm:
        print("进入async with语句块")
        res = await acm.do_something()
        print(res)

asyncio.run(main())
```

这个用的比较多，以异步操作数据库为例，由于数据库的连接需要耗时，`__aenter__()`方法中可以进行数据库的连接，`__aexit__()`方法则断开数据库连接。这样一来，我们的注意力就可以只放在操作数据上了。

一般不用自己实现，了解即可。

## `uvloop`效率提升

`uvloop`是`asyncio`的事件循环的第三方替代方案，**一定程度上可以提高事件循环的效率**。

若要进行替换，需要先安装`uvloop`模块：

```python
pip install uvloop
```

在事件循环创建前执行：

```python
import asyncio
import uvloop
asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())
```

即可完成替换，在后面编写正常的`asyncio`代码即可。在调用`asyncio.run()`时，内部的事件循环会自动变为`uvloop`。

很多性能不错的`Python`框架都使用了`asgi`中的`uvicorn`，它就是使用了`uvloop`替换了默认的事件循环。
